### ISSUE TITLE
Cleanups for 1.4.0 release

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#306]
- [x] Move into merging through pull requests
- [x] Check fastint downgrade handling of yield values
- [x] Update index page LOC, lowmemory startup size, code footprint (with buffer support) - #485
- [x] Accept plain buffer as a typed array constructor initializer? - #484
- [x] Buffer zeroing behavior with DUK_USE_ZERO_BUFFER_DATA unset? - #484
- [x] Typing for `duk_push_buffer_object()`, accept a buffer object as a base value, see discussion https://github.com/svaarala/duktape/issues/190 - #483, not sure if will be merged but tracked in that pull
- [x] -Wcast-qual warnings for DDDEBUG build
- [x] Buffer behavior changes? => No more for 1.4.0
- [x] Release notes: buffer behavior changes
- [x] MSVC /W4 fixes - #514

**Comment by svaarala** [comment]
Overall I'm planning the 1.4.0 release to be smaller than 1.3.0 which had a lot of changes. A good release interval would be around 3 months I think. Duktape 1.1 -> 1.2 took 3 months, 1.2 -> 1.3 took 5 months.

I'll be doing some release grooming and planning next week. I'll add a Wiki page for a roadmap overview so that this is a bit more transparent.

**Comment by svaarala** [comment]
Had to prune a few more things from the 1.4.0 release - there's a nice chunk of performance work in the release already, and I'd like to finish the call handling improvements. Moved other performance stuff like reimplementing the string table, further dispatch optimization etc, to 1.5.0.

Rough goal would be to finish actual feature development by end of November so that the release could be early December.

**Comment by fatcerberus** [comment]
Hm, I think I'm done with debugger features for the moment.  The only other one would be the heap walking for 1.5.0, but that's a little out of my league right now. :)

**Comment by fatcerberus** [comment]
1.4.0 is coming together pretty quickly it seems. :)

**Comment by svaarala** [comment]
I've been sidetracked this week. I'll probably need to postpone some performance / call handling related work to ensure the release gets done before the holidays. There's also a few large pending merges I'd like to start working into master, but only after the 1.4.0 release is done :)

**Comment by svaarala** [comment]
Code commits for 1.4.0 are now done, I'll be re-running the performance tests and doing compilation tests etc. If something is broken in master, now is a good time to bring it up :)

**Comment by fatcerberus** [comment]
Code freeze, nice.

1.4 turned out to be a pretty big improvement.  New debugger features, better error messages, strong finalizer guarantee, very noticeable performance gains... What's not to like?

**Comment by fatcerberus** [comment]
@svaarala
I'll pull down the latest master and run through some manual testing in minisphere to see if anything is acting oddly.  If so I'll let you know. :)

**Comment by JoshEngebretson** [comment]
@svaarala Awesome, I'll prioritize getting us updated to duktape 1.4 :+1:

**Comment by fatcerberus** [comment]
MSVC with `/W4` produces the following warnings (note: `/W3`, the generally recommended level, is silent):

```
1>duk_api_stack.c(2530): warning C4245: '=': conversion from 'int' to 'duk_small_uint_t', signed/unsigned mismatch
1>duk_api_stack.c(3752): warning C4245: '=': conversion from 'duk_int16_t' to 'duk_uint16_t', signed/unsigned mismatch
1>duk_api_stack.c(3948): warning C4127: conditional expression is constant
1>duk_bi_buffer.c(869): warning C4244: '=': conversion from 'duk_small_uint_t' to 'duk_uint8_t', possible loss of data
1>duk_bi_buffer.c(870): warning C4244: '=': conversion from 'duk_small_uint_t' to 'duk_uint8_t', possible loss of data
1>duk_bi_buffer.c(973): warning C4244: '=': conversion from 'duk_small_uint_t' to 'duk_uint8_t', possible loss of data
1>duk_bi_buffer.c(974): warning C4244: '=': conversion from 'duk_small_uint_t' to 'duk_uint8_t', possible loss of data
1>duk_hobject_props.c(4501): warning C4244: '=': conversion from 'duk_small_uint_t' to 'duk_uint8_t', possible loss of data
1>duk_hobject_props.c(4547): warning C4244: '=': conversion from 'duk_small_uint_t' to 'duk_uint8_t', possible loss of data
1>duk_hobject_props.c(5168): warning C4244: '=': conversion from 'duk_small_uint_t' to 'duk_uint8_t', possible loss of data
1>duk_hobject_props.c(5219): warning C4244: '=': conversion from 'duk_small_uint_t' to 'duk_uint8_t', possible loss of data
1>duk_hobject_props.c(5546): warning C4244: '=': conversion from 'duk_small_uint_t' to 'duk_uint8_t', possible loss of data
1>duk_js_compiler.c(923): warning C4244: '=': conversion from 'duk_reg_t' to 'duk_uint16_t', possible loss of data
1>duk_js_compiler.c(924): warning C4244: '=': conversion from 'duk_uint32_t' to 'duk_uint16_t', possible loss of data
1>duk_js_executor.c(4397): warning C4244: 'function': conversion from 'duk_uint_fast_t' to 'duk_small_uint_fast_t', possible loss of data
1>duk_js_var.c(1711): warning C4244: '=': conversion from 'duk_small_int_t' to 'duk_uint8_t', possible loss of data
1>duk_numconv.c(1186): warning C4244: '=': conversion from 'duk_small_int_t' to 'duk_uint8_t', possible loss of data
```

**Comment by svaarala** [comment]
Thanks, I'll fix those (at least some are present in 1.3.x too, I think).

**Comment by svaarala** [comment]
This one is difficult to fix:

```
1>duk_api_stack.c(3948): warning C4127: conditional expression is constant
```

It's caused by a sizeof() comparison which is indeed constant but depends on the target:

```
if (sizeof(duk_size_t) != sizeof(duk_uint_t)) {
    /* ... */
}
```

So the "then" block is only needed for some platforms, and on others it's intended to be optimized away completely.

Unfortunately there doesn't seem to be a way to do this more cleanly. Except assigning the sizeof() values to volatile variables and comparing them ;)

**Comment by svaarala** [comment]
@fatcerberus Does #514 fix the warnings (except for the "conditional expression is constant") for you?

**Comment by creationix** [comment]
@svaarala for the conditional, can you not use `#if .., #endif` in the preprocessor since it is a compile-time constant?

**Comment by svaarala** [comment]
I don't think `sizeof()` works in preprocessor conditionals?

I could add explicit sizeof macros for the types but that'd mean more type defines/wrappers in `duk_config.h` which I'd like to avoid, or compare the maximum values (although, interestingly, some of the integer limits provided by standard headers are not straight constants).

**Comment by fatcerberus** [comment]
@svaarala Yes, the warnings are fixed now.  I already said that in #514 of course, just posting here for posterity's sake.

**Comment by creationix** [comment]
@svaarala I see, that's unfortunate.  Googling solutions I found some pretty hacky stuff to create static asserts, but not the use case we have here.  Sorry for the noise.

**Comment by svaarala** [comment]
@creationix Yeah, it's quite difficult to squeeze the last warnings out :)

**Comment by JoshEngebretson** [comment]
Our latest dev snapshot is running Duktape 1.4 (latest master commit, right?) without issue :) Kudos and thanks for all the amazing work!

**Comment by fatcerberus** [comment]
Well, technically it's pre-1.4. ;-)
